File: 프로그래머스/2/42586. 기능개발/기능개발.py
'''
1. 아이디어
- 앞에 있는 기능이 먼저 수행 > 뒤에도 가능
    1) 각 작업 별로 남은 기간을 큐에 넣는다
    2) 큐에서 순서대로 꺼내면서,
        다음 순서의 작업이 먼저 끝났다면, count +1
        다음 순서의 작업이 더 늦게 끝난다면, append, new count +1
2. 시간복잡도
- 각 작업별로 남은 기간 구하기: O(N)
- 각 작업을 하나씩 꺼내면서 남은 기간 확인하기: O(N)
3. 변수
- 큐: 작업 순서대로 남은 기간 넣을 자료구조
'''

from collections import deque
import math

def solution(progresses, speeds):
    
    progress_time = deque()
    
    for i in range(len(progresses)):
        partTime = math.ceil((100 - progresses[i]) / speeds[i])
        progress_time.append(partTime)
        
    
    result = []
    
    while len(progress_time) >0 :
        first_prog = progress_time.popleft()
        count = 1
        while progress_time and first_prog >= progress_time[0]:
            progress_time.popleft()
            count += 1
        result.append(count)
        

    return list(result)

File: 프로그래머스/2/42586. 기능개발/기능개발.java
/*
[아이디어]
1. progressess에 들어있는 작업이 끝날 때까지 속도를 내서 배포를 해야 함.
2. Queue를 사용해서, 현재 가장 먼저 들어온 작업의 완료 일자보다 다음 작업들이 더 일찍 끝났으면 count+1 - O(N)
*/
import java.util.*; 
class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        
        List<Integer> list = new ArrayList<>();
        
        // 작업이 완료됐는지 확인하기 위한 자료구조 큐 q
        Queue<Integer> q = new ArrayDeque<>();
        
        for(int i=0; i<progresses.length; i++){
            int n = (100 - progresses[i]) / speeds[i];
            int m = (100 - progresses[i]) % speeds[i];
            int days = m > 0 ? n+1 : n;
            q.offer(days);
        }
        
        // 모든 작업이 완료될 때까지 반복
        while(!q.isEmpty()){
            int cur = q.poll();
            int next = !q.isEmpty() ? q.peek() : 100;
            int count = 1;
            
            while(cur >= next && !q.isEmpty()){
                q.poll();
                count++;
                if(!q.isEmpty()){
                    next = q.peek();
                }                
            };
            
            list.add(count);
            
        }
        
        int[] result = new int[list.size()];
        for(int i=0; i<list.size(); i++){
            result[i] = list.get(i);
        }
        
        return result;

    }
}